core/client/src/main/java/alluxio/client/block/stream/BlockInStream.java

```diff
@@ -0,0 +1,188 @@
+import alluxio.wire.LockBlockResult;
+import alluxio.wire.WorkerNetAddress;
+import alluxio.worker.block.io.LocalFileBlockReader;
+
+import com.google.common.io.Closer;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.InetSocketAddress;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Provides a stream API to read a block from Alluxio. An instance extending this class can be
+ * obtained by calling {@link AlluxioBlockStore#getInStream}. Multiple
+ * {@link BlockInStream}s can be opened for a block.
+ *
+ * This class provides the same methods as a Java {@link InputStream} with additional methods from
+ * Alluxio Stream interfaces.
+ */
+@NotThreadSafe
+public final class BlockInStream extends FilterInputStream implements BoundedStream, Seekable,
+    PositionedReadable, Locatable {
+  /** Helper to manage closeables. */
+  private final Closer mCloser;
+  private final BlockWorkerClient mBlockWorkerClient;
+  private final long mBlockId;
+  private final boolean mLocal;
+  private final PacketInStream mInputStream;
+
+  /**
+   * Creates an instance of local {@link BlockInStream} that reads from local worker.
+   *
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if it fails to create an instance
+   * @return the {@link BlockInStream} created
+   */
+  public static BlockInStream createLocalBlockInStream(long blockId, long blockSize,
+      WorkerNetAddress workerNetAddress, BlockStoreContext context, InStreamOptions options)
+      throws IOException {
+    Closer closer = Closer.create();
+
+    try {
+      BlockWorkerClient blockWorkerClient =
+          closer.register(context.createWorkerClient(workerNetAddress));
+      LockBlockResult lockBlockResult = blockWorkerClient.lockBlock(blockId);
+      LocalFileBlockReader localFileBlockReader =
+          new LocalFileBlockReader(lockBlockResult.getBlockPath());
+      PacketReader.Factory factory = new LocalFilePacketReader.Factory(localFileBlockReader);
+      return new BlockInStream(factory, blockId, blockSize, blockWorkerClient, options);
+    } catch (AlluxioException e) {
+      closer.close();
+      throw new IOException(e);
+    } catch (IOException e) {
+      closer.close();
+      throw e;
+    }
+  }
+
+  /**
+   * Creates an instance of remote {@link BlockInStream} that reads from a remote worker.
+   *
+   * @param blockId the block ID
+   * @param blockSize the block size
+   * @param workerNetAddress the worker network address
+   * @param context the block store context
+   * @param options the options
+   * @throws IOException if it fails to create an instance
+   * @return the {@link BlockInStream} created
+   */
+  public static BlockInStream createRemoteBlockInStream(long blockId, long blockSize,
```

Comment:
i'll leave it in the current way unless someone feels strongly about it because: 

Factory class is necessary if we need to pass the Factory object around in which it carries some state used to create the object. Otherwise, there is not much point to have a class to do the object creation.